h/h_api/bulk_api/command_processor.py

```diff
@@ -0,0 +1,186 @@
+        trigger the commands to be flushed to the executor.
+        """
+
+        if self.config is None:
+            raise CommandSequenceError("Not configured yet")
+
+        with self.batch.add(command):
+            # If we have any id references like `"id": {"$ref": ...}` we need
+            # to fill these out before we pass them to the executor. We do this
+            # now to get the earliest warning if we have any id references
+            # which don't match up
+            self.id_refs.fill_out(command.body.raw)
+
+    def _check_command_count(self, final=False):
+        """Check the command count matches expectations.
+
+        :param final: This is the final count check, not incremental
+        """
+
+        if not self.config:
+            return
+
+        total = self.config.total_instructions
+
+        if not final:
+            if self.command_count > total:
+                raise InvalidDeclarationError(
+                    f"More instructions ({self.command_count}) received than declared ({total})"
+                )
+
+        elif self.command_count != total:
+            raise InvalidDeclarationError(
+                f"Expected more instructions. Found {self.command_count} expected {total}"
+            )
+
+    def _execute_batch(self, command_type, data_type, batch):
+        """Prepare and execute a batch of commands.
+
+        This is passed to the `CommandBatch` object which will call us back
+        here when a batch is ready.
+        """
+
+        # Get configuration for this combo of command and data type
+        default_config = self.config.defaults_for(command_type, data_type)
+
+        # Prep commands to be sent to the executor
+        #
+        # All items in batch are the same type. We can use the first one
+        # to process the items in place. This will effect any command type
+        # specific commands which can be done for the executor
+        batch[0].prepare_for_execute(batch, default_config)
+
+        # We expect the executor to return a mapping from custom id references
+        # to concrete ids for any items with `$anchor`
+        reports = self.executor.execute_batch(
+            command_type=command_type,
+            data_type=data_type,
+            default_config=default_config,
+            batch=batch,
+        )
+
+        self._process_reports(data_type, batch, reports)
+
+    def _process_reports(self, data_type, batch, reports):
+        """
+        Store reports and update id references returned by the executor.
+
+        :param data_type: The data type these references are for
+        :param batch: The batch of commands containing id references
+        :param reports: A list of Report objects
+        """
+        if not isinstance(reports, list) or any(
+            not isinstance(item, Report) for item in reports
+        ):
+            raise TypeError(f"Expected a list of Report objects not: {reports}")
```

Comment:
Outside of development, this shouldn't really be happening, but it's a nice think to prevent anyway